server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java

```diff
@@ -1435,23 +1439,56 @@ private void closeConsistencyCheck() {
-          + tabletMetadata.getFilesMap() + " " + getDatafileManager().getDatafileSizes();
-      log.error(msg);
-    }
-  }
+  /**
+   * Checks that tablet metadata from the metadata table matches what this tablet has in memory. The
+   * caller of this method must acquire the updateCounter parameter before acquiring the
+   * tabletMetadata.
+   *
+   * @param updateCounter
+   *          used to check for conucurrent updates in which case this check is a no-op. See
+   *          {@link #getUpdateCount()}
+   * @param tabletMetadata
+   *          the metadata for this tablet that was acquired from the metadata table.
+   */
+  public synchronized void compareTabletInfo(MetadataUpdateCount updateCounter,
+      TabletMetadata tabletMetadata) {
+
+    // verify the given counter is for this tablet, if this check fail it indicates a bug in the
+    // calling code
+    Preconditions.checkArgument(updateCounter.getExtent().equals(getExtent()),
+        "Counter had unexpected extent %s != %s", updateCounter.getExtent(), getExtent());
+
+    // verify the given tablet metadata is for this tablet, if this check fail it indicates a bug in
+    // the calling code
+    Preconditions.checkArgument(tabletMetadata.getExtent().equals(getExtent()),
+        "Tablet metadata had unexpected extent %s != %s", tabletMetadata.getExtent(), getExtent());
 
-  public synchronized void compareTabletInfo(Long updateCounter, TabletMetadata tabletMetadata) {
     if (isClosed() || isClosing()) {
+      log.trace("AMCC Tablet {} was closed, so skipping check", tabletMetadata.getExtent());
       return;
     }
-    // if the counter didn't change, compare metadata to what is in memory
-    if (updateCounter == this.getUpdateCount()) {
-      this.compareToDataInMemory(tabletMetadata);
+
+    var dataFileSizes = getDatafileManager().getDatafileSizes();
+
+    if (!tabletMetadata.getFilesMap().equals(dataFileSizes)) {
+      // The counters are modified outside of locks before and after tablet metadata operations and
+      // data file updates so its very important to aquire the 2nd counts after doing the equality
+      // check above. If the counts are the same (as the ones acquired before reading metadata
+      // table) after the equality check above then we know the tablet did not do any updates while
+      // we were reading metadata and then comparing.
+      var latestCount = this.getUpdateCount();
+      if (updateCounter.overlapsUpdate() || !updateCounter.equals(latestCount)) {
+        log.trace(
+            "AMCC Tablet {} may have been updating its metadata while it was being read for check, so skipping check {} {}",
+            tabletMetadata.getExtent(), updateCounter, latestCount);
+      } else {
+        log.error("AMCC Data files in {} differ from in-memory data {} {} {} {}", extent,
+            tabletMetadata.getFilesMap(), dataFileSizes, updateCounter, latestCount);
```

Comment:
```suggestion
        log.error(
            "AMCC Data files in {} differ from in-memory data {}. dataFileSize: {} updateCounter: {} latestCount: {}",
            extent, tabletMetadata.getFilesMap(), dataFileSizes, updateCounter, latestCount);
```

Reply:
I am not sure what "AMCC" stands for but I guess it's OK to add uniqueness to the log statement. I just don't want to confuse users.

Reply:
AMCC was for Accumulo Metadata Consistency Check.  I can add a comment for that.  Do you think it should be removed?  I added because I like to be able to easily grep for all related messages.

Reply:
I like the ability to grep as well. I think a comment above where it gets logged would be nice. If a user is curious and they go to the code then they will see what it means.

Reply:
Added a comment in e60b6bb9b99a0783b4dbcd3aadb882123efed04c

Reply:
Jargon in the logs is very confusing. People won't know to grep for this specifically, and even if they go to the code to figure it out, it doesn't seem like it should be necessary. I'm not a fan of non-obvious jargon in log messages. It's also just as easy to grep for `'Data files in.*differ from in-memory'` or similar, if you know what you're expecting. If the jargon is still desired, I think it would be better to spell it out explicitly, even if it makes the log message longer.

Reply:
Maybe it could be put in a class called "AccumuloMetadataConsistencyCheck" or just use a named logger with a name like that.

Reply:
I think its fine for logging if it was shorthand for some type of name. I wouldn't consider it jargon if it was an acronym or an abbreviation of something.

Reply:
> Maybe it could be put in a class called "AccumuloMetadataConsistencyCheck" or just use a named logger with a name like that.

I don't want to refactor the code like that just for the log messages.  I removed the AMCC prefix from the error message and left it on the trace messages in 2296b47.  My thinking is that the error message is unsolicited, and the prefix may be confusing, however when someone is turning on tracing they are soliciting as much information as possible and I think this extra information may be useful (can be used for selection, deselection, tabulation, etc) even if a bit confusing at first.

Reply:
Seems like a reasonable compromise.